### ISSUE TITLE
Finding the maximum of the keys of a dictionary is really slow

### ISSUE THREAD

**Issue by Datseris** [JuliaDynamics/Agents.jl#303]
Ha, pop quiz time @Libbum  and @kavir1698 . In the entire forest fire model, which operation would you think takes the most amount of time? I thought it would be finding nearest neighbors.

Guess what:
```
model, agent_step!, model_step! = Models.forest_fire()
a = random_agent(model)
@btime move_agent!($a, $model);
# 54.379 ns (0 allocations: 0 bytes)

@btime space_neighbors($a, $model);
# 395.495 ns (5 allocations: 352 bytes) for forest fire with graph

@btime maximum(keys($model.agents))
# 209.300 μs (0 allocations: 0 bytes)

f(model) = maximum(collect(keys(model.agents)))
@btime f(model)
# 54.299 μs (3 allocations: 62.78 KiB)

@btime add_agent!($model, true)
# 209.799 μs (1 allocation: 48 bytes)
```

So, the **overwhelming majority of time** in our forest fire model is spend in calling `nextid`, which is done at every `add_agent!` call.

This is INSANE and I'll fix it in a new PR, which will modify the `ABM` struct to have a field `maxid` and this field is replaced whenever an agent with new id is added to the model.

**Comment by Datseris** [comment]
(btw my new space type decreases finding nearest neighbor speed about 5fold, but of course doesn't matter because the time of `step!` is unafected since `add_agent` takes 3 orders of magnitude more time)

**Comment by Libbum** [comment]
Oh geez. We are going to uncover some embarrassments in this project aren't we!

**Comment by Datseris** [comment]
you bet

**Comment by kavir1698** [comment]
Good catch!

**Comment by Datseris** [comment]
after my fix:
```julia
julia> @btime add_agent!($model, true)
  141.969 ns (1 allocation: 48 bytes)
```

**Comment by Libbum** [comment]
Closed via #304
